source/MaterialXRuntime/RtCommandEngine.h

```diff
@@ -0,0 +1,159 @@
+    RtCommandResult(bool success, const string& message = "") :
+        _success(success),
+        _message(message)
+    {
+    }
+
+    /// Construct a command result from a return object and optionally
+    /// a boolean state and an error message.
+    RtCommandResult(const RtObject& object, bool success = true, const string& message = "") :
+        _success(success),
+        _message(message),
+        _object(object)
+    {
+    }
+
+    /// Return true if the command executed succesfully.
+    bool success() const
+    {
+        return _success;
+    }
+
+    /// Return true if the command executed succesfully.
+    explicit operator bool() const
+    {
+        return success();
+    }
+
+    /// Return an error message if set when executing the command.
+    const string& getMessage() const
+    {
+        return _message;
+    }
+
+    /// Return an object resulting from executing the command.
+    const RtObject& getObject() const
+    {
+        return _object;
+    }
+
+private:
+    bool _success;
+    string _message;
+    RtObject _object;
+};
+
+/// @class RtCommandBase
+/// Base class for runtime commands.
+class RtCommandBase
+{
+public:
+    /// Destructor.
+    virtual ~RtCommandBase() {};
+
+    /// Execute the command.
+    virtual void execute(RtCommandResult& result) = 0;
+
+    /// Undo the command.
+    virtual void undo(RtCommandResult& result) = 0;
+
+    /// Redo the command.
+    virtual void redo(RtCommandResult& result)
+    {
+        execute(result);
+    }
+};
+
+/// A shared pointer to a runtime command.
+using RtCommandPtr = RtSharedPtr<RtCommandBase>;
+
+/// @class RtBatchCommand
+/// Class for executing batches of multiple commands.
+class RtBatchCommand : public RtCommandBase
+{
+public:
+    /// Constructor.
+    RtBatchCommand();
+
+    /// Destructor.
+    ~RtBatchCommand();
+
+    /// Add a command to the batch.
+    void addCommand(RtCommandPtr cmd);
+
+    /// Clear all commands in the batch.
+    void clearCommands();
+
+    /// Execute the command.
+    void execute(RtCommandResult& result) override;
+
+    /// Undo the command.
+    void undo(RtCommandResult& result)  override;
+
+    /// Redo the command.
+    void redo(RtCommandResult& result)  override;
+
+private:
+    void* _ptr;
+};
+
+/// @class RtCommandEngine
+/// Class handling runtime command execution.
+class RtCommandEngine
+{
+public:
+    /// Constructor.
+    RtCommandEngine();
+
+    /// Destructor.
+    ~RtCommandEngine();
+
+    /// Execute a new command.
+    void execute(RtCommandPtr cmd, RtCommandResult& result);
+
+    /// Execute the command on top of the undo queue.
+    void undo(RtCommandResult& result);
+
+    /// Execute the command on top of the redo queue.
+    void redo(RtCommandResult& result);
+
+    /// Flush the undo and redo queues.
+    /// All commands previously executed will no longer be undoable.
+    void flushUndoQueue();
```

Comment:
Do we need an enable/disable undo as well for control?
I don't think we want everything to be undoable all the time?

Reply:
Good point, that might be useful. We can still do any operation with the stage and prim level API though, which is immediate with less overhead and not undo-able. So maybe we can wait until we see a real use-case for disabling the undo queue.